plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoConnectorTest.java

```diff
@@ -338,6 +339,120 @@ public void testPredicatePushdown(String value)
+
+            MongoCollection<Document> collection = client.getDatabase(getSession().getSchema().orElseThrow()).getCollection(table.getName());
+            collection.insertMany(ImmutableList.of(document1, document2, document3, document4));
+
+            assertThat(query("SELECT id FROM %s WHERE id %2$s 1 OR discount %2$s 10".formatted(table.getName(), operator)))
+                    .isFullyPushedDown();
+
+            assertThat(query("SELECT id FROM %s WHERE id %2$s 1 OR 10 %2$s discount".formatted(table.getName(), operator)))
+                    .isFullyPushedDown();
+
+            assertThat(query("SELECT id FROM %s WHERE id %2$s 1 OR id %2$s discount".formatted(table.getName(), operator)))
+                    .isFullyPushedDown();
+
+            assertThat(query("SELECT id FROM %s WHERE id %2$s 1 OR discount %2$s id".formatted(table.getName(), operator)))
+                    .isFullyPushedDown();
+
+            assertThat(query("SELECT id FROM %s WHERE id %2$s 1 OR discount %2$s null".formatted(table.getName(), operator)))
+                    .isFullyPushedDown();
+
+            assertThat(query("SELECT id FROM %s WHERE id %2$s 1 OR null %2$s discount".formatted(table.getName(), operator)))
+                    .isFullyPushedDown();
+
+            assertThat(query("SELECT id FROM %s WHERE id %s 1 OR discount IS NULL".formatted(table.getName(), operator)))
+                    .isFullyPushedDown();
+
+            assertThat(query("SELECT id FROM %s WHERE id %s 1 OR discount IS NOT NULL".formatted(table.getName(), operator)))
+                    .isFullyPushedDown();
+        }
+    }
+
+    @Test(dataProvider = "comparisonOperator")
+    public void testComplexExpressionPredicatePushdownWithMismatchedDataType(String operator)
+    {
+        try (TestTable table = new TestTable(
+                getQueryRunner()::execute,
+                "test_complex_expression_predicate_pushdown_with_mismatched_data_type",
+                "(id INT, discount INT)")) {
+            Document document1 = new Document("id", 1).append("discount", 8);
+            Document document2 = new Document("id", 2).append("discount", "9");
+            Document document3 = new Document("id", 3).append("discount", 11);
+            Document document4 = new Document("id", 4).append("discount", "16");
+
+            MongoCollection<Document> collection = client.getDatabase(getSession().getSchema().orElseThrow()).getCollection(table.getName());
+            collection.insertMany(ImmutableList.of(document1, document2, document3, document4));
+
+            if (operator.equals("!=") || operator.equals(">") || operator.equals(">=")) {
+                assertThat(query("SELECT id FROM %s WHERE id = 1 OR discount %s 10".formatted(table.getName(), operator)))
+                        .matches("VALUES 1, 2, 3, 4"); // TODO Enabling complex filter pushdown returns the mismatched datatype document as well
```

Comment:
But the same query with complex filter pushdown disabled returns a different result right ? I think with complex filter enabled we either need to throw an exception - so user could re-run the same query with the config disabled or we shouldn't allow pushdown of those operation via complex expression pushdown.